pkg/config/config.go

```diff
@@ -97,123 +90,43 @@ func (b Bundles) Options() (res [][]bundles.BundleOption) {
 	return
 }
 
-func (c Config) Unmarshal(o interface{}) error {
-	return yaml.Unmarshal([]byte(c.String()), o)
-}
-
-func (c Config) Data() map[string]interface{} {
-	return c.originalData
-}
-
-func (c Config) String() string {
-	if len(c.originalData) == 0 {
-		dat, err := yaml.Marshal(c)
-		if err == nil {
-			return string(dat)
-		}
-	}
-
-	dat, _ := yaml.Marshal(c.originalData)
-	if c.header != "" {
-		return AddHeader(c.header, string(dat))
-	}
-	return string(dat)
-}
-
-func (c Config) Query(s string) (res string, err error) {
-	s = fmt.Sprintf(".%s", s)
-	jsondata := map[string]interface{}{}
-
-	// c.String() takes the original data map[string]interface{} and Marshals into YAML, then here we unmarshall it again?
-	// we should be able to use c.originalData and copy it to jsondata
-	err = yaml.Unmarshal([]byte(c.String()), &jsondata)
-	if err != nil {
-		return
-	}
-	query, err := gojq.Parse(s)
-	if err != nil {
-		return res, err
-	}
-
-	iter := query.Run(jsondata) // or query.RunWithContext
-	for {
-		v, ok := iter.Next()
-		if !ok {
-			break
-		}
-		if err, ok := v.(error); ok {
-			return res, fmt.Errorf("failed parsing, error: %w", err)
-		}
-
-		dat, err := yaml.Marshal(v)
-		if err != nil {
-			break
-		}
-		res += string(dat)
-	}
-	return
-}
-
 // HasConfigURL returns true if ConfigURL has been set and false if it's empty.
 func (c Config) HasConfigURL() bool {
 	return c.ConfigURL != ""
 }
 
-func allFiles(dir []string) []string {
-	files := []string{}
-	for _, d := range dir {
-		if f, err := listFiles(d); err == nil {
-			files = append(files, f...)
-		}
-	}
-	return files
-}
+func Scan(opts ...collector.Option) (c *Config, err error) {
+	result := &Config{}
 
-func Scan(opts ...Option) (c *Config, err error) {
-	o := &Options{}
+	o := &collector.Options{}
 	if err := o.Apply(opts...); err != nil {
-		return nil, err
+		return result, err
 	}
 
-	c = parseConfig(o.ScanDir, o.NoLogs)
-
-	if o.MergeBootCMDLine {
-		d, err := machine.DotToYAML(o.BootCMDLineFile)
-		if err == nil { // best-effort
-			yaml.Unmarshal(d, c) //nolint:errcheck
-			// Merge back to originalData only config which are part of the config structure
-			// This avoid garbage as unrelated bootargs to be merged in.
-			dat, err := yaml.Marshal(c)
-			if err == nil {
-				yaml.Unmarshal(dat, &c.originalData) //nolint:errcheck
-			}
-		}
-	}
+	genericConfig, err := collector.Scan(o)
+	if err != nil {
+		return result, err
 
-	if c.HasConfigURL() {
-		err = c.fetchRemoteConfig()
-		if !o.NoLogs && err != nil {
-			fmt.Printf("WARNING: Couldn't fetch config_url: %s\n", err.Error())
```

Comment:
This is no longer a soft error (we used to ignore non-accessible remote configurations). This is why we have a failed test in CI.

I wonder if it's good practice to not fail when configuration is missing. Let's say, a user sets up something really important security-wise, in a remote configuration. It's possible that by ignoring the error, we end up with a less secure system (or even one that doesn't even work). Ignored errors are rarely a good thing and result in broken systems that are hard to debug.

My opinion is we should keep it as it is in this PR and change the test. @mudler thoughts?